main/src/main/scala/sbt/internal/server/BuildServerProtocol.scala

```diff
@@ -92,36 +97,54 @@ object BuildServerProtocol {
       }
     },
     bspEnabled := true,
-    bspWorkspace := bspWorkspaceSetting.value,
+    bspSbtEnabled := true,
+    bspFullWorkspace := bspFullWorkspaceSetting.value,
+    bspWorkspace := bspFullWorkspace.value.scopes,
     bspWorkspaceBuildTargets := Def.taskDyn {
-      val workspace = Keys.bspWorkspace.value
+      val workspace = Keys.bspFullWorkspace.value
       val state = Keys.state.value
-      val allTargets = ScopeFilter.in(workspace.values.toSeq)
+      val allTargets = ScopeFilter.in(workspace.scopes.values.toSeq)
+      val sbtTargets: List[Def.Initialize[Task[BuildTarget]]] = workspace.builds.map {
+        case (buildTargetIdentifier, loadedBuildUnit) =>
+          val buildFor = workspace.buildToScope.getOrElse(buildTargetIdentifier, Nil)
+          sbtBuildTarget(loadedBuildUnit, buildTargetIdentifier, buildFor)
+      }.toList
       Def.task {
         val buildTargets = Keys.bspBuildTarget.all(allTargets).value.toVector
-        state.respondEvent(WorkspaceBuildTargetsResult(buildTargets))
-        buildTargets
+        val allBuildTargets = buildTargets ++ sbtTargets.join.value
+        state.respondEvent(WorkspaceBuildTargetsResult(allBuildTargets))
+        allBuildTargets
       }
     }.value,
     // https://github.com/build-server-protocol/build-server-protocol/blob/master/docs/specification.md#build-target-sources-request
     bspBuildTargetSources := Def.inputTaskDyn {
       val s = state.value
-      val workspace = bspWorkspace.value
       val targets = spaceDelimited().parsed.map(uri => BuildTargetIdentifier(URI.create(uri)))
-      val filter = ScopeFilter.in(targets.map(workspace))
+      val workspace = bspFullWorkspace.value.filter(targets)
+      val filter = ScopeFilter.in(workspace.scopes.values.toList)
       // run the worker task concurrently
       Def.task {
         val items = bspBuildTargetSourcesItem.all(filter).value
-        val result = SourcesResult(items.toVector)
+        val buildItems = workspace.builds.toVector.map {
+          case (id, loadedBuildUnit) =>
+            val base = loadedBuildUnit.localBase
+            val sbtFiles = configurationSources(base)
+            val defDir = projectStandard(base)
```

Comment:
That's not precise enough because the build client would think that the `project/plugins.sbt` and all the files in `project/project/` directories are source files of the main build target, which is wrong.

Also the `project/src/main/scala` belongs to the sbt build target. And the user is even free to add any other source directory with:
```scala
// project/build.sbt
unmanagedSourceDirectories += baseDirectory.value / "custom"
```
Or even he/she can add some generated sources.

Maybe, a good way to fix this is to pass the all sources from the meta-build to the build context through the `PluginData`.

Reply:
In other words, we should put more information in here so that the BSP build target is consistent with the meta-build definition:
https://github.com/sbt/sbt/blob/9274a3e365d3cb826ede31598afde412ab6625c1/main/src/main/scala/sbt/internal/Load.scala#L1163-L1174.

Reply:
WIP: https://github.com/sbt/sbt/compare/develop...retronym:topic/bsp-import-plugin-data?expand=1

Reply:
I've added a commit to this PR that implements your suggestion.

I don't go as far as trying to represent the meta-build (or meta-meta-...-builds!) (`project/*.sbt` / `project/project/*.scala`) as a build target. This _is_ supported by the BSP spec but not yet supported by IntelliJ.